### PR TITLE
Adding new overland flow formulations

### PR DESCRIPTION
**Pull request by lecondon**
- Added three new boundary conditions as modules - OverlandKinematic, OverlandDiffusive and Seepage. 
- OverlandKinematic is similar to the original OverlandFlow boundary condition but uses a slightly modified flux formulation that uses the slope magnitude and it is developed to use face centered slopes (as opposed to grid centered) and does the upwinding internally. 
- Added documentation for these BCs to the manual and developed new test cases - overland_slopingslab_DWE.tcl, overland_slopingslab_KWE.tcl, overland_tiltedv_DWE.tcl, overland_tiltedV_KWE.tcl, Overland_FlatICP.tcl
- Two new options were also added to the terrain following grid formulation to be consistent with the upwinding approach used in the new overland flow formulation these are specified with the new TFGUpwindFormullation keys documented in the manual 
- For both OverlandDiffusive and OverlandKinematic analytical jacobians were implemented in the new modules and these were tested and can be verified in the new test cases noted above. 
- The new BCs were instantiated using the modified approach developed by MCB in problem_bc.h

### CODE REVIEW

**Comment by lecondon** [comment]
Additional notes on the macro update by MB
- Replaced magic number system based on magic string indexing into named macros defined under bc_pressure.h
- Moved Type0 and BCPressureType0 (and Type1 and etc) struct declarations under an X-Macro table defined in bc_pressure.h, instead of split between bc_pressure.h and bc_pressure_package.c
- Added descriptive macro names to BCPressurePackageNewPublicXtra control flow for defining new, what used to be "Type0", structs, as well as allocation, storage, and accessor macros for them
- Documented the process for adding new BC structs and types at the top of bc_pressure.h, including how to branch on new boundary conditions in the looping macros (i.e. adding new BC types for new equations)

**Comment by smithsg84** [comment]
The regression suite isn’t passing on this pull request for the autoconf build.

https://travis-ci.org/parflow/parflow/builds/566196696?utm_source=github_status&utm_medium=notification

My guess is script doesn’t handle multiple PASSING comments in a single test script.  But there are also missing files:

***************************************************************************
Running overland_FlatICP.tcl
couldn't read file "overland_FlatICP.tcl": no such file or directory
*****************************************************************************

**Comment by reedmaxwell** [comment]
We added new tests and more tests pass than the autoconf make check counts (so it gets a -33 or something and “fails”)— when I pull I see that file, I think it has a capital “o” not lower case.  i.e. Overland_FlatCP.tcl

Reed

Reed M. Maxwell, Ph.D.
Rowlinson Professor of Hydrology
Department of Geology and Geological Engineering
Director, Integrated GroundWater Modeling Center
Colorado School of Mines
http://inside.mines.edu/~rmaxwell/

On Aug 5, 2019, at 3:16 PM, Steven Smith <notifications@github.com<mailto:notifications@github.com>> wrote:


The regression suite isn’t passing on this pull request for the autoconf build.

https://travis-ci.org/parflow/parflow/builds/566196696?utm_source=github_status&utm_medium=notification

My guess is script doesn’t handle multiple PASSING comments in a single test script. But there are also missing files:

________________________________

Running overland_FlatICP.tcl
couldn't read file "overland_FlatICP.tcl": no such file or directory

________________________________

**Comment by lecondon** [comment]
I fixed the file name issue and I also added the new tests to the cmake list and they appear to be passing now. The only issue is that with the cmake approach it just shows each of the new tests passing once but there are multiple tests within each of the new scripts. How can I specify to have these report out separately?

**Comment by smithsg84** [comment]
The Ctest system is counting each executable as a single test.    I’m not finding many solutions. The one common suggestion is to invoke the same executable with an argument that indicates the test. 

https://baoilleach.blogspot.com/2013/06/using-ctest-with-multiple-tests-in.html

In this case the tcl file would take an extra argument for the specific test to execute

tclsh test_script.tcl 1 1 1 test_name1
tclsh test_script.tcl 1 1 1 test_name2

Assuming script already takes in topology as arguments.   

Run only the indicated test, possibly using a TCL switch statement to select the test.

File: pfsimulator/parflow_lib/richards_jacobian_eval.c [hunk 1]
```diff
@@ -333,6 +337,11 @@ void    RichardsJacobianEval(
             public_xtra->type = overland_flow;
           }
           break;
+///          case 9:
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/richards_jacobian_eval.c]: Remove comment code

File: pfsimulator/parflow_lib/richards_jacobian_eval.c [hunk 2]
```diff
@@ -333,6 +337,11 @@ void    RichardsJacobianEval(
             public_xtra->type = overland_flow;
           }
           break;
+///          case 9:
+///          {
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/richards_jacobian_eval.c]: Remove commented code

File: pfsimulator/parflow_lib/richards_jacobian_eval.c [hunk 3]
```diff
@@ -616,10 +625,45 @@ void    RichardsJacobianEval(
                     + rpp[ip + sz_v] * ddp[ip + sz_v];
 
       //@RMM
-      x_dir_g = Mean(gravity * sin(atan(x_ssl_dat[ioo])), gravity * sin(atan(x_ssl_dat[ioo + 1])));
-      x_dir_g_c = Mean(gravity * cos(atan(x_ssl_dat[ioo])), gravity * cos(atan(x_ssl_dat[ioo + 1])));
-      y_dir_g = Mean(gravity * sin(atan(y_ssl_dat[ioo])), gravity * sin(atan(y_ssl_dat[ioo + sy_v])));
-      y_dir_g_c = Mean(gravity * cos(atan(y_ssl_dat[ioo])), gravity * cos(atan(y_ssl_dat[ioo + sy_v])));
+//      x_dir_g = Mean(gravity * sin(atan(x_ssl_dat[ioo])), gravity * sin(atan(x_ssl_dat[ioo + 1])));
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/richards_jacobian_eval.c]: Remove commented code

File: pfsimulator/parflow_lib/richards_jacobian_eval.c [hunk 4]
```diff
@@ -54,13 +54,15 @@ enum JacobianType {
   no_nonlinear_jacobian,
   not_set,
   simple,
-  overland_flow
+  overland_flow //,
+//  seepage_face
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/richards_jacobian_eval.c]: Remove commented code

File: pfsimulator/parflow_lib/richards_jacobian_eval.c [hunk 5]
```diff
@@ -1123,21 +1175,33 @@ void    RichardsJacobianEval(
                     op = up;
                     prod_val = rpp[ip + sz_v] * den_d;
 
-                    lower_cond = (pp[ip]) - 0.5 * dz * Mean(z_mult_dat[ip], z_mult_dat[ip + sz_v]) * dp[ip] * gravity;
-                    upper_cond = (value) + 0.5 * dz * Mean(z_mult_dat[ip], z_mult_dat[ip + sz_v]) * den_d * gravity;
-                    diff = lower_cond - upper_cond;
+//                    lower_cond = (pp[ip]) - 0.5 * dz * Mean(z_mult_dat[ip], z_mult_dat[ip + sz_v]) * dp[ip] * gravity;
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/richards_jacobian_eval.c]: Remove old code

File: .gitignore [hunk 6]
```diff
@@ -1,5 +1,8 @@
 # Created by https://www.gitignore.io
 
+/test/*
+/examples/*
```
[the remark this answers is not in the document]
Comment (lecondon) [reply .gitignore]: Done

File: docs/manuals/files.tex [hunk 7]
```diff
@@ -3420,6 +3421,22 @@ \subsection{Code Parameters}
 pfset Solver.Drop 1E-6
 \end{verbatim}\end{display}
 
+\pfkey{double}{Solver.OverlandDiffusive.Epsilon}{1E-5}
+{
+This key provides a lower value for the $\bar{S_{f}}$ for the {\bf OverlandDiffusive} boundary condition.
```
[the remark this answers is not in the document]
Comment (lecondon) [reply docs/manuals/files.tex]: I changed this to say that it provides a minimum value. Hopefully that is more clear.

File: pfsimulator/parflow_lib/overlandflow_eval_Kin.c [hunk 8]
```diff
@@ -0,0 +1,379 @@
+/*BHEADER*********************************************************************
+ *
+ *  Copyright (c) 1995-2009, Lawrence Livermore National Security,
```
[the remark this answers is not in the document]
Comment (lecondon) [reply pfsimulator/parflow_lib/overlandflow_eval_Kin.c]: I removed everything before the GNU License lines

File: pfsimulator/parflow_lib/overlandflow_eval_Kin.c [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+*  This module computes the contributions for the spatial discretization of the
+*  kinematic wave approximation for the overland flow boundary condition:KE,KW,KN,KS.
+*
+*  It also computes the derivatives of these terms for inclusion in the Jacobian.
+
+* @LEC, @RMM
+*****************************************************************************/
+#include "parflow.h"
+#include "llnlmath.h"
+//#include "llnltyps.h"
```
[the remark this answers is not in the document]
Comment (lecondon) [reply pfsimulator/parflow_lib/overlandflow_eval_Kin.c]: Done

File: pfsimulator/parflow_lib/richards_jacobian_eval.c (hunk 1 above)
Comment (lecondon) [reply pfsimulator/parflow_lib/richards_jacobian_eval.c]: Done

File: pfsimulator/parflow_lib/richards_jacobian_eval.c (hunk 2 above)
Comment (lecondon) [reply pfsimulator/parflow_lib/richards_jacobian_eval.c]: Done

File: pfsimulator/parflow_lib/richards_jacobian_eval.c (hunk 3 above)
Comment (lecondon) [reply pfsimulator/parflow_lib/richards_jacobian_eval.c]: Done

File: pfsimulator/parflow_lib/richards_jacobian_eval.c (hunk 4 above)
Comment (lecondon) [reply pfsimulator/parflow_lib/richards_jacobian_eval.c]: Done

File: pfsimulator/parflow_lib/richards_jacobian_eval.c (hunk 5 above)
Comment (lecondon) [reply pfsimulator/parflow_lib/richards_jacobian_eval.c]: Done

File: pfsimulator/parflow_lib/bc_pressure.h [hunk 10]
```diff
@@ -10,64 +10,193 @@
...
+    BC_TYPE(OverlandKinematic, {                    \
+        double *values;                             \
+      })                                            \
+    BC_TYPE(OverlandDiffusive, {                    \
+          double *values;                           \
+        })
+/*----------------------------------------------------------------
+ * Constructor, getter, and setter for Type structs in BCPressurePackage
+ *----------------------------------------------------------------*/
+#define NewTypeStruct(type, var)                  \
```
[the remark this answers is not in the document]
Comment (KoenigKrote) [reply pfsimulator/parflow_lib/bc_pressure.h]: Added

File: pfsimulator/parflow_lib/bc_pressure.h [hunk 11]
```diff
@@ -202,4 +341,53 @@ typedef struct {
  * BCPressure Data constants used in the program.
  *--------------------------------------------------------------------------*/
 
+
+/*----------------------------------------------------------------
+ * BCPressurePackage Actions
+ *----------------------------------------------------------------*/
+#define InputType(public_xtra, i)               \
```
[the remark this answers is not in the document]
Comment (KoenigKrote) [reply pfsimulator/parflow_lib/bc_pressure.h]: Added

**Pull request merged**